portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -46,29 +59,89 @@
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    int maxComments = getMaxCommentsToReturn(request);
-
+    JsonObject json = new JsonObject();
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      String name = getUserName(userService.getCurrentUser().getUserId());
+
+      // If the user has logged in for the first time, set their name as their Google account nickname.
+      if (name == null) {
+        name = userService.getCurrentUser().getNickname();
+        NameServlet.upsertUserInfo(name);
+      }
+
+      String urlToRedirectToAfterUserLogsOut = "/";
+      String logoutUrl = userService.createLogoutURL(urlToRedirectToAfterUserLogsOut);
+      String userEmail = userService.getCurrentUser().getEmail();
+
+      json.addProperty("logoutUrl", logoutUrl);
+      json.addProperty("email", userEmail);
+      json.addProperty("name", name);
+    } else {
+      // Add a login link to the response if the user is not logged in.
+      String urlToRedirectToAfterUserLogsIn = "/";
+      String loginUrl = userService.createLoginURL(urlToRedirectToAfterUserLogsIn);
+      json.addProperty("loginUrl", loginUrl);
+    }
+    
     // Query up to maxComments comment entities from Datastore with the user's specified sorting option.
+    int maxComments = getMaxCommentsToReturn(request);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     Query query = createCommentQuery(request);
     List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(maxComments));
 
-    // Construct a list of comments from the queried entities.
-    List<Comment> comments = results.stream().map((entity) -> {
-      long id = entity.getKey().getId();
-      String author = (String) entity.getProperty("author");
-      String commentText = (String) entity.getProperty("commentText");
-      long timestamp = (long) entity.getProperty("timestamp");
+    // Check if there are comments to translate.
+    if (results.isEmpty()) {
+      json.add("comments", new JsonArray());
+    } else {
+      // Construct a stream of comment texts from the queried entities.
+      Stream<String> commentTexts = results.stream().map((entity) -> {
+        String commentText = (String) entity.getProperty("commentText");
+        return commentText;
```

Comment:
If the function within map() is only one line, you don't need brackets or the return statement. It could be shortened to:

entity -> (String) entity.getProperty("commentText")